### PR TITLE
Fixed `gauss_kl` when the shape of the Gaussian parameters is unknown

### PR DESCRIPTION
**Pull request by rhaps0dy**
Also added a test to prevent future regressions.

This addresses the bug in both #734 and #725  .

### CODE REVIEW

File: tests/test_kldiv.py [hunk 1]
```diff
@@ -145,5 +145,15 @@ def test_oned(session_tf, white, mu, sqrt, K_batch):
     np.testing.assert_allclose(kl.eval(), kl_tf.eval())
 
 
+def test_unknown_size_inputs(session_tf):
+    mu_ph = tf.placeholder(settings.float_type, [None, None])
+    sqrt_ph = tf.placeholder(settings.float_type, [None, None, None])
+    mu = np.ones([1, 4], dtype=settings.float_type)
```
Comment (awav) [inline tests/test_kldiv.py]: np.ones => tf.ones

File: tests/test_kldiv.py [hunk 2]
```diff
@@ -145,5 +145,15 @@ def test_oned(session_tf, white, mu, sqrt, K_batch):
...
+    sqrt = np.ones([4, 1, 1], dtype=settings.float_type)
+    np.testing.assert_allclose(
+        session_tf.run(gauss_kl(*map(tf.constant, [mu, sqrt]))),
```
Comment (awav) [inline tests/test_kldiv.py]: no need in `map` anymore after using tf.ones

File: tests/test_kldiv.py [hunk 3]
```diff
@@ -145,5 +145,15 @@ def test_oned(session_tf, white, mu, sqrt, K_batch):
     np.testing.assert_allclose(kl.eval(), kl_tf.eval())
 
 
+def test_unknown_size_inputs(session_tf):
```
Comment (awav) [inline tests/test_kldiv.py]: it would be nice to have docstring describing the issue and reference to github issue

File: tests/test_kldiv.py [hunk 4]
```diff
@@ -145,5 +145,15 @@ def test_oned(session_tf, white, mu, sqrt, K_batch):
...
+    mu_ph = tf.placeholder(settings.float_type, [None, None])
+    sqrt_ph = tf.placeholder(settings.float_type, [None, None, None])
+    mu = np.ones([1, 4], dtype=settings.float_type)
+    sqrt = np.ones([4, 1, 1], dtype=settings.float_type)
+    np.testing.assert_allclose(
+        session_tf.run(gauss_kl(*map(tf.constant, [mu, sqrt]))),
+        session_tf.run(gauss_kl(mu_ph, sqrt_ph), {mu_ph: mu, sqrt_ph: sqrt}))
```
Comment (awav) [inline tests/test_kldiv.py]: > Signature: sess.run(fetches, feed_dict=None, options=None, run_metadata=None)

Therefore your call should be
```
feed_dict = {mu_ph: mu, sqrt_ph: sqrt}
session_tf.run(gauss_kl(mu_ph, sqrt_ph), feed_dict=feed_dict))
```

**Comment by awav** [comment]
Hello, @rhaps0dy! I would appreciate if you could summarize why this issue happened and how you solved it.

PS: from pull request template,
> Description
> Describe your contribution by writing down the context and reasons. Make it clear whether it is a bug > fix, feature or enhancement. Do not make it very wordy, keep it short and simple.
>
> Bad example: This is a bug-fix found at #111.
> Bad example: Add new SuperMagic optimizer.
> Good example: Working on feature #111, we found that parameters of VGP are not instantiated > properly. It causes arbitrary failures at training. The bug is fixed by adding necessary parameters to the list of trainables.

File: tests/test_kldiv.py (hunk 1 above)
Comment (rhaps0dy) [reply tests/test_kldiv.py]: Does that work, even for passing them to the placeholders? (I tested and it seems that it doesn't). The point of having np.ones is to be able to have unknown-size tensors later.

**Comment by rhaps0dy** [comment]
Hi @awav. Sorry, I thought it would be fine having done the bug description carefully in #725. I'll just copy the issue, it contains all the information required by the pull request template.

Point taken about the docstring though. I've also run pylint, and the only warnings I introduced are these:

```
W:148,29: Redefining name 'session_tf' from outer scope (line 25) (redefined-outer-name)
W:156, 4: Redefining name 'mu' from outer scope (line 52) (redefined-outer-name)
W:157, 4: Redefining name 'sqrt' from outer scope (line 68) (redefined-outer-name)
```

but the rest of the file is littered with them as well.

# `gauss_kl` error when mean has non-static shape

When running `gpflow.kullback_leiblers.gauss_kl` on something with shape undefined at graph-building time, such as a `tf.placeholder`, the following happens:
```
TypeError: Value passed to parameter 'out_type' has DataType float64 not in list of allowed values: int32, int64
```
And [this GPflow line](https://github.com/GPflow/GPflow/blob/master/gpflow/kullback_leiblers.py#L73) is the one that causes the error. It seems that tensorflow has different code paths for fully defined and partially defined shapes, see [this Tensorflow snippet](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/array_ops.py#L399-L404).

Minimum failing example:
```python
import gpflow
import tensorflow as tf
import numpy as np

mu = tf.ones([1, 4], dtype=gpflow.settings.float_type)
sig = tf.ones([4, 1, 1], dtype=gpflow.settings.float_type)
mu_ph = tf.placeholder(gpflow.settings.float_type, [1, None])

with tf.Session() as sess:
    print(sess.run(gpflow.kullback_leiblers.gauss_kl(mu, sig)))  # works
    print(sess.run(gpflow.kullback_leiblers.gauss_kl(mu_ph, sig),
                   feed_dict={mu_ph: np.ones([1, 4])}))  # doesn't work
```

File: tests/test_kldiv.py [hunk 5]
```diff
@@ -145,5 +145,21 @@ def test_oned(session_tf, white, mu, sqrt, K_batch):
...
+    one unknown parameter, `gauss_kl` would blow up. This happened because
+    `tf.size` can only output types `tf.int32` or `tf.int64`.
+    """
+    mu_ph = tf.placeholder(settings.float_type, [None, None])
+    sqrt_ph = tf.placeholder(settings.float_type, [None, None, None])
+    mu = np.ones([1, 4], dtype=settings.float_type)
+    sqrt = np.ones([4, 1, 1], dtype=settings.float_type)
+    feed_dict = {mu_ph: mu, sqrt_ph: sqrt}
+    np.testing.assert_allclose(
+        session_tf.run(gauss_kl(*map(tf.constant, [mu, sqrt]))),
```
Comment (awav) [inline tests/test_kldiv.py]: *One remark:* it is always preferable to be more explicit in tests and more wordy.

```
known_shape_tf = gauss_kl(*map(tf.constant, [mu, sqrt]))
unknown_shape_tf = gauss_kl(mu_ph, sqrt_ph)
known_shape = session_tf.run(known_shape_tf)
unknown_shape = session_tf.run(unknown_shape_tf, , feed_dict=feed_dict)
np.testing.assert_allclose(known_shape, unknown_shape)
```

**Comment by rhaps0dy** [comment]
Thank you very much for finishing this!

**Pull request merged**